### PR TITLE
Use a DFS to compare CTFE snapshots

### PR DESCRIPTION
**Pull request by ecstatic-morse**
This is an alternative fix to #52475 that hopes to address some parts of #54384.

The major development in this PR is the use of an actual depth-first search to compare the heap between snapshots, instead of the current recursive method. We treat the set of pointers on the stack as an ordered list of "roots". Using this as the initial search stack, a DFS can enumerate all reachable memory in a predictable order. This is important, since without a stable ordering, we would need to compute an isomorphism for the allocation graph of each snapshot to compare them. 

Equality between snapshots is checked by iterating in lock-step over their allocation graphs and comparing each `Allocation` using an `IgnoreAllocId` wrapper with custom `PartialEq` and `Hash` impls. If the allocations are equivalent but for `AllocId`s, all pointers in the allocation are mapped to a DFS index and then compared.

I find the existing implementation pretty inscrutable, so I'm not equipped to speculate on performance differences. At the very least, my method will use less memory when a collision *does* occur, since we no longer clone `Memory` in full, just the reachable portions.

Some aspects of #54384 remain unaddressed, most notably what to do when we fail to resolve an `AllocId`. My PR takes what I believe is the same approach as the existing implementation; it converts the `InterpError` into a `None` and continues traversing the allocation graph. I don't know what the correct behavior would be.

r? @RalfJung 
cc @oli-obk

### CODE REVIEW

File: src/librustc/mir/interpret/pointer.rs [hunk 1]
```diff
@@ -202,6 +202,13 @@ impl<'tcx, Tag> Pointer<Tag> {
         Pointer { alloc_id: self.alloc_id, offset: self.offset, tag: () }
     }
 
+    /// Erase any allocation IDs.
+    ///
+    /// Used when comparing heap snapshots.
+    pub fn erase_id(self) -> Pointer<Tag, ()> {
```
Comment (RalfJung) [inline src/librustc/mir/interpret/pointer.rs]: Please rename to `erase_alloc_id`.

But isn't it easier to just say `ptr1.offset == ptr2.offset`? Pointers with `AllocId` type `()` are kind of pointless (all puns intended).

File: src/librustc_mir/interpret/operand.rs [hunk 2]
```diff
@@ -132,6 +144,36 @@ impl<Tag> Operand<Tag> {
 
         }
     }
+
+    /// Returns a list of all `AllocId`s present in this `Operand`.
+    pub(super) fn alloc_ids(&self) -> SmallVec<[AllocId; 2]>
```
Comment (RalfJung) [inline src/librustc_mir/interpret/operand.rs]: Could you move this into `snapshot.rs`? I surely hope nothing else will ever need this kind of operation.^^

File: src/librustc/mir/interpret/pointer.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc/mir/interpret/pointer.rs]: Oh, you are recursively doing through even for `Place`. But how can that make any sense, given that all AllocIds are considered equal at this point...?

File: src/librustc_mir/interpret/snapshot.rs [hunk 3]
```diff
@@ -5,27 +5,19 @@
...
-use super::{Frame, Memory, Operand, MemPlace, Place, Immediate, ScalarMaybeUndef, LocalValue};
+use rustc::mir::interpret::{Allocation, InterpResult};
+use rustc::ty::layout::Size;
+use rustc_data_structures::fx::{FxHasher, FxHashSet};
+
+use super::{memory, Frame, Memory};
 use crate::const_eval::CompileTimeInterpreter;
+use crate::interpret::{LocalState, LocalValue};
+
+type CtfeMemory<'mir, 'tcx> = Memory<'mir, 'tcx, CompileTimeInterpreter<'mir, 'tcx>>;
```
Comment (RalfJung) [inline src/librustc_mir/interpret/snapshot.rs]: This seems worth moving to `const_eval.rs`.

File: src/librustc_mir/const_eval.rs [hunk 4]
```diff
@@ -465,13 +465,16 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter<'mir,
...
-            *ecx.tcx,
-            span,
-            &ecx.memory,
-            &ecx.stack[..],
-        )
+        if ecx.machine.loop_detector.is_empty() {
+            // FIXME(#49980): make this warning a lint
+            ecx.tcx.sess.span_warn(span,
+                "Constant evaluating a complex constant, this might take some time");
+        }
```
Comment (RalfJung) [inline src/librustc_mir/const_eval.rs]: Why did you move this out? Seems like a layering violation for this code to care about "emptiness" of the loop detector.

Comment (RalfJung) [reply src/librustc_mir/const_eval.rs]: Oh I see, you could remove the `tcx` and `span` arguments then. That is a good argument, but `is_empty` is not a great name.

File: src/librustc_mir/interpret/snapshot.rs [hunk 5]
```diff
@@ -46,361 +38,338 @@ pub(crate) struct InfiniteLoopDetector<'mir, 'tcx> {
...
-
-                $struct_name {
-                    $( $field: __impl_snapshot_field!($field, __ctx $(, $delegate)?) ),*
-                }
+        // Copy all reachable allocations that exist in `memory.alloc_map` into a new `alloc_map`.
+        //
+        // FIXME: How should we handle allocations stored in `tcx`? If these can be mutable, we
+        // should be moving them into `alloc_map` here. Otherwise the loop detector may fire even
+        // though memory is changing. I suspect that everything in `tcx` is immutable during
+        // const-eval, however.
```
Comment (RalfJung) [inline src/librustc_mir/interpret/snapshot.rs]: Indeed, const-eval never mutates `tcx` allocations.

File: src/librustc_mir/interpret/snapshot.rs [hunk 6]
```diff
@@ -46,361 +38,338 @@ pub(crate) struct InfiniteLoopDetector<'mir, 'tcx> {
...
-    stmt: usize,
-}
-
-impl<'a, 'mir, 'tcx, Ctx> Snapshot<'a, Ctx> for &'a Frame<'mir, 'tcx>
-    where Ctx: SnapshotContext<'a>,
-{
-    type Item = FrameSnapshot<'a, 'tcx>;
+/// A wrapper type for the various interpreter data structures that compares them without
+/// considering the numeric value of `AllocId`s.
+struct AllocIdInvariant<'a, T>(&'a T);
```
Comment (RalfJung) [inline src/librustc_mir/interpret/snapshot.rs]: "invariant" is an odd term here, that indicates that something doesn't change.

`IgnoreAllocId` sounds more accurate? But I have no idea why it even makes any sense to do this.^^

**Comment by RalfJung** [comment]
> Some aspects of #54384 remain unaddressed, most notably what to do when we fail to resolve an AllocId. My PR takes what I believe is the same approach as the existing implementation; it converts the InterpError into a None and continues traversing the allocation graph. I don't know what the correct behavior would be.

This means we have a dangling pointer, right?

File: src/librustc_mir/interpret/memory.rs [hunk 7]
```diff
@@ -941,3 +923,56 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> Memory<'mir, 'tcx, M> {
...
+impl<M: Machine<'mir, 'tcx>> Iterator for DepthFirstSearch<'mem, 'mir, 'tcx, M> {
+    type Item = (AllocId, InterpResult<'tcx, &'mem Allocation<M::PointerTag, M::AllocExtra>>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let DepthFirstSearch { stack, visited, memory } = self;
+
+        let id = stack.pop()?;
+        let alloc = memory.get_raw(id);
+
+        if let Ok(alloc) = alloc {
```
Comment (RalfJung) [inline src/librustc_mir/interpret/memory.rs]: Is this where you are ignoring `InterpError`? This seems reasonable.

We should make sure that we don't catch errors requiring allocation though, similar to [this](https://github.com/rust-lang/rust/blob/f5c81e0a986e4285d3d0fd781a1bd475753eb12c/src/librustc_mir/transform/const_prop.rs#L391). Maybe create a new helper method at the error type for this?

File: src/librustc_mir/interpret/memory.rs [hunk 8]
```diff
@@ -114,24 +114,6 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> HasDataLayout for Memory<'mir, 'tcx, M>
     }
 }
 
-// FIXME: Really we shouldn't clone memory, ever. Snapshot machinery should instead
-// carefully copy only the reachable parts.
```
Comment (RalfJung) [inline src/librustc_mir/interpret/memory.rs]: I'm so happy to see this go away. :)

File: src/librustc/mir/interpret/pointer.rs (hunk 1 above)
Comment (ecstatic-morse) [reply src/librustc/mir/interpret/pointer.rs]: :+1: to `erase_alloc_id`. This is the part that makes `AllocId`s compare as equal. It means we can use the derived `PartialEq` impl. The alternative is to overload `AllocIdInvariant` for more types, which would also be fine.

File: src/librustc_mir/const_eval.rs (hunk 4 above)
Comment (ecstatic-morse) [reply src/librustc_mir/const_eval.rs]: `has_been_invoked`?

File: src/librustc_mir/interpret/snapshot.rs [hunk 9]
```diff
@@ -46,361 +38,338 @@ pub(crate) struct InfiniteLoopDetector<'mir, 'tcx> {
...
+        let live_allocs = self.live_allocs();
+        let other_live_allocs = other.live_allocs();
 
-        let all_bytes = 0..self.len();
-        // This 'inspect' is okay since following access respects undef and relocations. This does
-        // influence interpreter exeuction, but only to detect the error of cycles in evalution
-        // dependencies.
-        let bytes = self.inspect_with_undef_and_ptr_outside_interpreter(all_bytes);
+        live_allocs.map(|(_, alloc)| alloc.ok().map(AllocIdInvariant))
+            .eq(other_live_allocs.map(|(_, alloc)| alloc.ok().map(AllocIdInvariant)))
```
Comment (ecstatic-morse) [inline src/librustc_mir/interpret/snapshot.rs]: @RalfJung. I discard the contents of the `InterpError` with `ok`. Do you think we can differentiate between null pointers, pointers to ZSTs and dangling pointers by looking at `InterpError` like you mentioned above? Do ZST pointers have zero-sized backing allocations?

File: src/librustc_mir/interpret/snapshot.rs (hunk 6 above)
Comment (ecstatic-morse) [reply src/librustc_mir/interpret/snapshot.rs]: Yeah, I would not be opposed to reverting to the `AllocId`-sensitive method, or removing this entirely. It is pretty cool when you accidentally omit a `break` and you're told there's a provably infinite loop in your program, and I have some sentimental connection to this code since it was my first real rust PR. However, I think a configurable instruction limit (with a low default) coupled with a real lint would have basically the same effect.

Once loops are implemented, people will actually start hitting bugs/performance issues in this code. I wanted to make sure I understood what's going on now. I'm worried people will see long-running constant evaluations grind to a halt because of this machinery and have no way to turn it off.

cc @oli-obk

File: src/librustc/mir/interpret/pointer.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc/mir/interpret/pointer.rs]: But why does it make any sense to compare just the offsets?

File: src/librustc_mir/const_eval.rs (hunk 4 above)
Comment (RalfJung) [reply src/librustc_mir/const_eval.rs]: No that's negated... `is_unused`?

File: src/librustc/mir/interpret/pointer.rs (hunk 1 above)
Comment (ecstatic-morse) [reply src/librustc/mir/interpret/pointer.rs]: Other code is responsible for comparing the allocation that each `AllocId` references, specifically the `PartialEq` implementation of `InterpRef`. Am I misunderstanding the question?

File: src/librustc_mir/const_eval.rs (hunk 4 above)
Comment (ecstatic-morse) [reply src/librustc_mir/const_eval.rs]: `!has_been_invoked` is negated but `!is_unused` isn't?

**Review by oli-obk — COMMENTED** [review]
I started a review, and forgot about it. I'll continue it later. For now just dumping the little status it has

File: src/librustc_mir/const_eval.rs (hunk 4 above)
Comment (RalfJung) [reply src/librustc_mir/const_eval.rs]: > !has_been_invoked is negated but is_unused isn't?

Well I wanted to avoid the `!` is all. It's way too easy to miss.

File: src/librustc_mir/interpret/memory.rs (hunk 7 above)
Comment (ecstatic-morse) [reply src/librustc_mir/interpret/memory.rs]: I'm not sure what you mean by "errors requiring allocation". Are there certain errors that we should `bug` on instead of ignoring? Is `Ub` one of these?

**Comment by ecstatic-morse** [comment]
@oli-obk, @RalfJung the necessary changes have been made to handle memory cycles. I also added a (very wordy) high-level explanation of the new code in the module docs. Finally, I resolved many of the concerns, although I would still like to do something to address [this one](https://github.com/rust-lang/rust/pull/66946#discussion_r352467568).

File: src/librustc_mir/interpret/memory.rs (hunk 7 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/memory.rs]: Any error type that has a `String` or other heap types in their fields will allocate when being thrown, and the catching will then destroy the allocated values. If this allocate->throw->catch->destroy happens in a high frequency we're wasting a lot of cycles with all the allocating/deallocating.

So, the idea is to make sure that any error that is caught and acted upon (and not just rethrown) does not allocate

Comment (oli-obk) [reply src/librustc_mir/interpret/memory.rs]: Idea: Should we make the set of errors that `Memory` can emit a separate type (with appropriate `Into` impls) so we know the set of errors that `Memory` methods can emit and can make sure all of them don't allocate?

File: src/librustc_mir/interpret/snapshot.rs [hunk 10]
```diff
@@ -46,361 +85,423 @@ pub(crate) struct InfiniteLoopDetector<'mir, 'tcx> {
...
-    stack: Vec<Frame<'mir, 'tcx>>,
+impl Hash for IgnoreAllocId<'_, LocalState<'tcx>> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // See `PartialEq` impl above for description of ignored fields.
+        let super::LocalState {
+            value,
+            layout: _,
+        } = self.0;
+
+        match value {
```
Comment (oli-obk) [inline src/librustc_mir/interpret/snapshot.rs]: you also need to hash the discriminant when you hash an enum's fields, otherwise the `Live` variant may end up with the same hash as a `Dead` or `Uninitialized` variant (without being a true has collision).

File: src/librustc_mir/interpret/snapshot.rs [hunk 11]
```diff
@@ -1,31 +1,70 @@
...
+//! as a "snapshot".
+//!
+//! While comparing stacks between snapshots is straightforward, comparing heaps between snapshots
+//! is not. This is because `AllocId`s, which determine the target of a pointer, are frequently
+//! thrown away and recreated, even if the actual value in memory did not change. Consider the
+//! following code:
+//!
+//! ```rust,ignore(const-loop)
+//! const fn inf_loop() {
+//!     // Function call to prevent promotion.
```
[the remark this answers is not in the document]
Comment (ecstatic-morse) [reply src/librustc_mir/interpret/snapshot.rs]: Doh! Can't believe I missed this. My worry is that the test for [issue 52475](https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/const-eval/issue-52475.rs) isn't really creating new `AllocId`s, since the `&0` is getting promoted out.

File: src/librustc_mir/interpret/memory.rs (hunk 7 above)
Comment (ecstatic-morse) [reply src/librustc_mir/interpret/memory.rs]: That makes sense. I'll probably wait for someone else to do this? It feels like y'all have a decent idea of what you want here.

**Comment by RalfJung** [comment]
What is the long-term plan here? From other conversation, it seems like one plan was to remove the loop checker entirely. So we probably should finish that discussion before spending more time on this PR?

Also I got quite a backlog of PRs to handle so @oli-obk if you could take over review that would be much appreciated.

File: src/librustc_mir/interpret/snapshot.rs [hunk 12]
```diff
@@ -1,31 +1,66 @@
...
+//! Stack pointers:    x  x    (DFS starts with leftmost pointer on top of stack)
+//!                    |  |
+//! Heap allocs:     1 o  o 3
+//!                    | /|
+//!                    |/ |
+//!                  2 o  o 4
+//! ```
+//!
+//! Instead of comparing `AllocId`s between snapshots, we first map `AllocId`s in each snapshot to
+//! their DFS index, then compare those instead.
```
Comment (RalfJung) [inline src/librustc_mir/interpret/snapshot.rs]: That makes sense. However, what I would have expected then is to see `Pointer<Tag, DfsIndex>` instead of `Pointer<Tag, AllocId>`, and then one would compare those. What I still don't get is why we have a way to compare pointers *ignoring* the `AllocId`. That means we must have a magic scheme somewhere to figure out all the `AllocId` we ignored and compare their `DfsIndex` instead? That sounds quite round-about and fragile, on first sight. (I didn't read the code, though.)

**Comment by oli-obk** [comment]
> From other conversation, it seems like one plan was to remove the loop checker entirely. So we probably should finish that discussion before spending more time on this PR?

This was discussed in https://rust-lang.zulipchat.com/#narrow/stream/146212-t-compiler.2Fconst-eval/topic/Time.20limits.20during.20const-eval/near/183588492

and from the last comment

> I'm gonna open a PR removing the loop detector after #67260 gets merged.

I am assuming that we can close this PR
